### PR TITLE
Move Rasterizer::Draw's discard_callback to Delegate

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR `Rasterizer::Draw`'s `discard_callback` parameter, which is assigned by `Shell`, to `Delegate`, which is also implemented by `Shell`. This refactory makes the API cleaner.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/embedder/tests/embedder_metal_unittests.mm [hunk 1]
```diff
@@ -218,18 +218,22 @@ GrBackendTexture backend_texture(texture_size.width(), texture_size.height(), Gr
 
   auto rendered_scene = context.GetNextSceneImage();
 
+  printf("main 1\n");fflush(stdout);
```
Comment (chinmaygarde) [inline shell/platform/embedder/tests/embedder_metal_unittests.mm:221]: Spurious printfs.

**Review by chinmaygarde — APPROVED** [review]
Not sure if this was ready for review just yet but we ran into this in triage. LGTM after the nits.

File: shell/common/rasterizer.cc [hunk 2]
```diff
@@ -209,19 +208,22 @@ RasterStatus Rasterizer::Draw(
...
             std::move(item->frame_timings_recorder);
         float device_pixel_ratio = item->device_pixel_ratio;
-        if (discard_callback(view_id, *layer_tree.get())) {
+        if (weak_this &&
+            delegate.ShouldDiscardLayerTree(view_id, *layer_tree.get())) {
           raster_status = RasterStatus::kDiscarded;
         } else {
-          raster_status = DoDraw(std::move(frame_timings_recorder),
-                                 std::move(layer_tree), device_pixel_ratio);
+          raster_status =
```
Comment (dnfield) [inline shell/common/rasterizer.cc:224]: I think this should be `else if (weak_this)`

Comment (dnfield) [reply shell/common/rasterizer.cc:224]: You're potentially dereferencing null here.

File: shell/common/shell.h [hunk 3]
```diff
@@ -689,6 +689,10 @@ class Shell final : public PlatformView::Delegate,
   // |Rasterizer::Delegate|
   fml::TimePoint GetLatestFrameTargetTime() const override;
 
+  // |Rasterizer::Delegate|
+  bool ShouldDiscardLayerTree(int64_t view_id,
+                              flutter::LayerTree& tree) override;
```
Comment (loic-sharma) [inline shell/common/shell.h:694]: Can this be `const`?

```suggestion
                              const flutter::LayerTree& tree) override;
```

**Review by loic-sharma — APPROVED** [review]
LGMT

**Pull request merged**